### PR TITLE
Add profile support

### PR DESCRIPTION
**Pull request by rhjostone**
This is MR intends to address #147 through several enhancements:

* `CommandOptionParser` has two new optional parameters -- `default_profile` and `configuration_directory`
* `BaseClientCommandContainer` has been added which doesn't immediately parse configuration files, instead creating an empty container which can then be loaded once a profile has been chosen.

### CODE REVIEW

File: kobo/client/__init__.py
Comment (rohanpm) [inline kobo/client/__init__.py:100]: We should avoid adding a `load` method due to backwards-compatibility concerns.

This class is designed to be subclassed.  Unfortunately, adding new methods to any library class which is intended to be subclassed is a risky matter as you have no idea if someone using the library happens to have also implemented a method named "load" on their subclass.  If they have, then your call to `self.load` later calls *their* method and not this one, and that'll most likely crash.

This is only called from one place anyway, could we just have the self.conf.load_from* moved to ClientCommandContainer constructor?

File: kobo/client/__init__.py
Comment (rohanpm) [inline kobo/client/__init__.py:96]: Could we get a comment or doc string summarizing the difference between these two classes?  As it is, it's not clear what functionality should go into Base and what doesn't.

For example it's not clear why set_hub isn't in Base.

**Review by rohanpm — CHANGES_REQUESTED** [review]

**Review by rhjostone — COMMENTED** [review]
I've submitted some new changes for a follow-up review. Please let me know what you think!

**Comment by rohanpm** [comment]
I don't see anything objectionable here, I think it just needs to be covered by tests.

**Comment by rhjostone** [comment]
@rohanpm I've added some tests which appears to improve code coverage. Does anything else need to be improved?

File: kobo/client/__init__.py
Comment (rohanpm) [inline kobo/client/__init__.py:115]: Hmm, I think I mentioned earlier that super() with no args doesn't work on python2 (and you do want this to be compatible with python2).

The fact that CI is not failing seems to indicate that this code isn't covered by a test, and the coverage report confirms it:
https://coveralls.io/builds/35010530/source?filename=kobo%2Fclient%2F__init__.py#L115

Could you please ensure it's covered at least to the point of "does not crash"?

**Comment by rohanpm** [comment]
@rhjostone it's looking close.  Can I ask, have you tried making use of this code yet from Pub, to verify that it provides what's needed?

**Comment by rhjostone** [comment]
@rohanpm I've tested these changes against upstream pub for backwards compatibility, as well as with my own downstream patches. It seems to provide the functionality needed for #148.

File: kobo/client/__init__.py
Comment (rhjostone) [reply kobo/client/__init__.py]: @rohanpm I think these latest changes should address python2 compatibility.

**Comment by rohanpm** [comment]
@rhjostone it looks alright to me.

Would you mind doing a few things to finalize this:

- Rewrite the PR message so it no longer says it's a work in progress
- Reduce the number of commits in the PR, e.g. using "git rebase -i" to combine related commits to one.  I think it's reasonable for the entire PR to be just one commit.
- Could you write a sentence or two for the changelog?  You don't have to put them into the changelog yourself, you could just mention it here.  Basically, how do we summarize this change in a sentence or two for the users of kobo, in a similar style as is written in the changelog at https://github.com/release-engineering/kobo#changelog .

**Comment by rohanpm** [comment]
Also, please let me know if you see any reason this should not immediately be released once merged.

**Pull request merged**
